### ISSUE TITLE
UnimplementedError | InvalidArgumentError when running tf.nn.conv2d with data_format NCHW

### ISSUE THREAD

**Issue by bugpromax** [tensorflow/tensorflow#46623]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 2019
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.3.1
- Python version: 3.6
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
As declared in the specification, tf.nn.con2d should support both "NHWC" and "NCHW" format. However, per our experiment, running it with data in "NCHW" format will trigger some unexpected behaviors.

*data_format – An optional `string` from: `"NHWC", "NCHW"`. Defaults to `"NHWC"`. Specify the data format of the input and output data. With the default format "NHWC", the data is stored in the order of: [batch, height, width, channels]. Alternatively, the format could be "NCHW", the data storage order of: [batch, channels, height, width].*

**Code to reproduce **
x = np.random.randn(1, 3, 12, 12)
x_tf = tf.convert_to_tensor(x, dtype=tf.float32)

weights=np.ones([2,2,3,8])
weights_tf = tf.convert_to_tensor(weights, dtype=tf.float32)
stride=1
sess = tf.compat.v1.Session()
tf_conv2d = tf.nn.conv2d(x_tf,
                         weights_tf,
                         strides=[1, stride, stride, 1],
                         padding="SAME",data_format="NCHW",use_cudnn_on_gpu=False)
sess.run(tf.global_variables_initializer())
tf_result = sess.run(tf_conv2d)
print(tf_result.shape)

when set strides=[1,1,1,1]  trigger a bug:
"UnimplementedError (see above for traceback): Generic conv implementation only supports NHWC tensor format for now."

when set strides=[1,2,2,1]  trigger a bug:
"InvalidArgumentError (see above for traceback): Current implementation does not yet support strides in the batch and depth dimensions."

**Describe the expected behavior**
No exception during execution

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by Saduf2019** [comment]
@bugpromax 
I ran the code shared and face a different error, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/d941c83edf23fafc4a0e784904cbf23d/untitled507.ipynb). If possible please share a colab gist with the error reported.

**Comment by bugpromax** [comment]
the code should be:

import tensorflow as tf
import numpy as np

x = np.random.randn(1, 3, 12, 12)
x_tf = tf.convert_to_tensor(x, dtype=tf.float32)

weights=np.ones([2,2,3,8])
weights_tf = tf.convert_to_tensor(weights, dtype=tf.float32)
stride=1
sess = tf.compat.v1.Session()
tf_conv2d = tf.nn.conv2d(x_tf,
weights_tf,
strides=[1, stride, stride, 1],
padding="SAME",data_format="NCHW")
sess.run(tf.global_variables_initializer())
tf_result = sess.run(tf_conv2d)
print(tf_result.shape)

**Comment by bugpromax** [comment]
please find the [gist here](https://colab.research.google.com/drive/1iRdEG_6mp9WQb5PXaN1eErP9w-9vx6tT?usp=sharing)

**Comment by bugpromax** [comment]
import tensorflow as tf
import numpy as np

x = np.random.randn(1, 3, 12, 12)
x_tf = tf.convert_to_tensor(x, dtype=tf.float32)

weights=np.ones([2,2,3,8])
weights_tf = tf.convert_to_tensor(weights, dtype=tf.float32)
stride=1
sess = tf.compat.v1.Session()
tf_conv2d = tf.nn.conv2d(x_tf,
weights_tf,
strides=[1, stride, stride, 1],
padding="SAME",data_format="NCHW")
sess.run(tf.global_variables_initializer())
tf_result = sess.run(tf_conv2d)
print(tf_result.shape)

**Comment by Saduf2019** [comment]
I am able to replicate the error reported on tf 2.4 and nightly , please find the [gist here](https://colab.research.google.com/gist/Saduf2019/90593711239b0d59afbb6293385c4186/untitled513.ipynb)

**Comment by bugpromax** [comment]
May I know how everything is going with this issue

**Comment by saikumarchalla** [comment]
Issue still exists in TF 2.6 Nightly.

**Comment by chunduriv** [comment]
@bugpromax, Sorry for late response. 

TensorFlow does not support `NCHW` on `CPU` . For more information please refer [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/backend.py#L912-L915). 

I tried to run your code on Colab using `TF 2.8` in eager mode and it is working. Please find the [gist](https://colab.research.google.com/gist/chunduriv/bf2d19e29ccbfc784e9426cbf9461f29/46623.ipynb) here for reference.Thanks!
